Foundation/GTMSignalHandler.m

```diff
@@ -67,20 +50,34 @@ - (id)initWithSignal:(int)signo
       return nil;
     }
 
-    signo_ = signo;
-    target_ = target;  // Don't retain since target will most likely retain us.
-    action_ = action;
-    GTMAssertSelectorNilOrImplementedWithArguments(target_,
-                                                   action_,
+    GTMAssertSelectorNilOrImplementedWithArguments(target,
+                                                   action,
                                                    @encode(int),
                                                    NULL);
 
-    // We're handling this signal via kqueue, so turn off the usual signal
+    // We're handling this signal via libdispatch, so turn off the usual signal
     // handling.
-    signal(signo_, SIG_IGN);
+    if (signal(signo, SIG_IGN) == SIG_ERR) {
+      _GTMDevLog(@"could not ignore signal %d.  Errno %d", signo, errno);  // COV_NF_LINE
+    }
 
     if (action != NULL) {
-      [self registerWithKQueue];
+      signalSource_ = dispatch_source_create(DISPATCH_SOURCE_TYPE_SIGNAL,
+                                             signo,
+                                             0,
+                                             dispatch_get_main_queue());
+      dispatch_source_set_event_handler(signalSource_, ^(void) {
+        NSMethodSignature *methodSig
+            = [target methodSignatureForSelector:action];
+        _GTMDevAssert(methodSig != nil, @"failed to get the signature?");
+        NSInvocation *invocation
+            = [NSInvocation invocationWithMethodSignature:methodSig];
+        [invocation setTarget:target];
+        [invocation setSelector:action];
+        [invocation setArgument:(void*)&signo atIndex:2];
```

Comment:
This will be a local for the arg, and not the address of the original stack param, right?

Reply:
which should be fine correct?